README.md
# Sensirion Embedded I2C SDP Driver

This is a generic embedded driver for the Sensirion SDP8xx and SDP3x sensor series. It enables developers to communicate
with the SDP8xx and SDP3x sensors on different hardware platforms by only adapting the I2C communication related source
files.

<center><img src="images/SDP8xx.png" width="250px"><img src="images/SDP3x.png" width="250px"></center>

# Getting started

## Implement the I2C Interface

So we need to adjust two files according to your platform.

### Edit `sensirion_i2c_hal.c`

This file contains the implementation of the sensor communication, which
depends on your hardware platform. We provide function stubs for your
hardware's own implementation.
Sample implementations are available for some platforms:
[`sample-implementations`](sample-implementations). For Linux based platforms
like Raspberry Pi you can just replace the unimplemented HAL template with the
implementation in `sample-implementations/linux_user_space/`:

```
cp sample-implementations/linux_user_space/sensirion_i2c_hal.c ./
```

### Edit `sensirion_config.h`

Skip this part for Linux based platforms since everything is already setup for
this case.

Otherwise you need to check if the libraries `<stdint.h>` and `<stdlib.h>` are
provided by your toolchain, compiler or system. If you have no idea on how to
do that you can skip this step for now and come back when you get errors
related to these names when compiling the driver.
The features we use from those libraries are type definitions for integer sizes
from `<stdint.h>` and `NULL` from `<stdlib.h>`. If they are not available you
need to specify the following integer types yourself:

* `int64_t` = signed 64bit integer
* `uint64_t` = unsigned 64bit integer
* `int32_t` = signed 32bit integer
* `uint32_t` = unsigned 32bit integer
* `int16_t` = signed 16bit integer
* `uint16_t` = unsigned 16bit integer
* `int8_t` = signed 8bit integer
* `uint8_t` = unsigned 8bit integer

In addition to that you will need to specify `NULL`. For both we have a
detailed template where you just need to fill in your system specific values.

Now we are ready to compile and run the example usage for your sensor.

## Compile and Run

Pass the source `.c` and header `.h` files in this folder into your C compiler
and run the resulting binary. This step may vary, depending on your platform.
Here we demonstrate the procedure for Linux based platforms:

1. Open up a terminal.
2. Navigate to the directory where this README is located.
3. Make sure the correct i2c address is configured in sdp_i2c.h (SDP_I2C_ADDRESS)
4. Run `make` (this compiles the example code into one executable binary).
5. Run the compiled executable with `./sdp_i2c_example_usage`
6. Now you should see the first measurement values appear in your terminal. As
   a next step you can adjust the example usage file or write your own main
   function to use the sensor.

# Background

## Files

### sensirion\_i2c.[ch]

In these files you can find the implementation of the I2C protocol used by
Sensirion sensors. The functions in these files are used by the embedded driver
to build the correct frame out of data to be sent to the sensor or receive a
frame of data from the sensor and convert it back to data readable by your
machine. The functions in here calculate and check CRCs, reorder bytes for
different byte orders and build the correct formatted frame for your sensor.

### sensirion\_i2c\_hal.[ch]

These files contain the implementation of the hardware abstraction layer used
by Sensirion's I2C embedded drivers. This part of the code is specific to the
underlying hardware platform. This is an unimplemented template for the user to
implement. In the `sample-implementations/` folder we provide implementations
for the most common platforms.

### sensirion\_config.h

In this file we keep all the included libraries for our drivers and global
defines. Next to `sensirion_i2c_hal.c` *it's the only file you should need to
edit to get your driver working.*

### sensirion\_common.[ch]

In these files you can find some helper functions used by Sensirion's embedded
drivers. It mostly contains byte order conversions for different variable
types. These functions are also used by the UART embedded drivers therefore
they are kept in their own file.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [0.1.0] - 2021-08-31

Initial release

[0.1.0]: https://github.com/Sensirion/embedded-i2c-sdp/releases/tag/0.1.0

tests/sdp_i2c_test.cpp
/*
 * THIS FILE IS AUTOMATICALLY GENERATED
 *
 * I2C-Generator: 0.2.0
 * Yaml Version: 0.1.0
 * Template Version: 0.7.0-32-g5b8f0ff
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "i2c_mux.h"
#include "sdp_i2c.h"
#include "sensirion_common.h"
#include "sensirion_i2c.h"
#include "sensirion_i2c_hal.h"
#include "sensirion_test_setup.h"
#include <inttypes.h>
#include <stdio.h>

TEST_GROUP (SDP_Tests) {
    void setup() {
        sensirion_i2c_hal_init();

        sdp_stop_continuous_measurement();
        // Select MUX 1 channel 1 (TODO: attach SDP to testbed and adjust mux
        int16_t error = sensirion_i2c_mux_set_single_channel(0x71, 1);
        CHECK_EQUAL_ZERO_TEXT(error, "sensirion_i2c_mux_set_single_channel")
    }

    void teardown() {
        sensirion_i2c_hal_free();
    }
};

TEST (
    SDP_Tests,
    SDP_Test_start_continuous_measurement_with_mass_flow_t_comp_and_averaging) {
    int16_t error;
    error =
        sdp_start_continuous_measurement_with_mass_flow_t_comp_and_averaging();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_start_continuous_measurement_with_"
                                 "mass_flow_t_comp_and_averaging");
    error = sdp_stop_continuous_measurement();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_stop_continuous_measurement");
}

TEST (SDP_Tests, SDP_Test_start_continuous_measurement_with_mass_flow_t_comp) {
    int16_t error;
    error = sdp_start_continuous_measurement_with_mass_flow_t_comp();
    CHECK_EQUAL_ZERO_TEXT(
        error, "sdp_start_continuous_measurement_with_mass_flow_t_comp");
    error = sdp_stop_continuous_measurement();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_stop_continuous_measurement");
}

TEST (
    SDP_Tests,
    SDP_Test_start_continuous_measurement_with_diff_pressure_t_comp_and_averaging) {
    int16_t error;
    error =
        sdp_start_continuous_measurement_with_diff_pressure_t_comp_and_averaging();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_start_continuous_measurement_with_"
                                 "diff_pressure_t_comp_and_averaging");
    error = sdp_stop_continuous_measurement();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_stop_continuous_measurement");
}

TEST (SDP_Tests,
      SDP_Test_start_continuous_measurement_with_diff_pressure_t_comp) {
    int16_t error;
    error = sdp_start_continuous_measurement_with_diff_pressure_t_comp();
    CHECK_EQUAL_ZERO_TEXT(
        error, "sdp_start_continuous_measurement_with_diff_pressure_t_comp");
    error = sdp_stop_continuous_measurement();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_stop_continuous_measurement");
}

TEST (SDP_Tests, SDP_Test_stop_continuous_measurement) {
    int16_t error;
    error = sdp_stop_continuous_measurement();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_stop_continuous_measurement");
}

TEST (SDP_Tests,
      SDP_Test_trigger_measurement_with_mass_flow_t_comp_and_averaging) {
    int16_t error;
    error = sdp_trigger_measurement_with_mass_flow_t_comp_and_averaging();
    CHECK_EQUAL_ZERO_TEXT(
        error, "sdp_trigger_measurement_with_mass_flow_t_comp_and_averaging");
}

TEST (SDP_Tests, SDP_Test_trigger_measurement_with_diff_pressure_t_comp) {
    int16_t error;
    error = sdp_trigger_measurement_with_diff_pressure_t_comp();
    CHECK_EQUAL_ZERO_TEXT(error,
                          "sdp_trigger_measurement_with_diff_pressure_t_comp");
}

TEST (SDP_Tests, SDP_Test_read_measurement) {
    int16_t error;
    int16_t differential_pressure;
    int16_t temperature;
    int16_t scaling_factor;
    error =
        sdp_start_continuous_measurement_with_diff_pressure_t_comp_and_averaging();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_start_continuous_measurement_with_"
                                 "diff_pressure_t_comp_and_averaging");

    error = sdp_read_measurement(&differential_pressure, &temperature,
                                 &scaling_factor);
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_read_measurement");
    printf("Differential pressure: %i\n", differential_pressure);
    printf("Temperature: %i\n", temperature);
    printf("Scaling factor: %i\n", scaling_factor);

    error = sdp_stop_continuous_measurement();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_stop_continuous_measurement");
}

TEST (SDP_Tests, SDP_Test_read_measurement_single_shot) {
    int16_t error;
    int16_t differential_pressure;
    int16_t temperature;
    int16_t scaling_factor;
    error = sdp_trigger_measurement_with_diff_pressure_t_comp();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_start_continuous_measurement_with_"
                                 "diff_pressure_t_comp_and_averaging");

    error = sdp_read_measurement(&differential_pressure, &temperature,
                                 &scaling_factor);
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_read_measurement");
    printf("Differential pressure: %i\n", differential_pressure);
    printf("Temperature: %i\n", temperature);
    printf("Scaling factor: %i\n", scaling_factor);
}

TEST (SDP_Tests, SDP_Test_sleep_mode) {
    int16_t error;
    error = sdp_enter_sleep_mode();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_enter_sleep_mode");
    error = sdp_exit_sleep_mode();
    CHECK_EQUAL_TEXT(-1, error, "sdp_exit_sleep_mode");
    error = sdp_exit_sleep_mode();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_exit_sleep_mode");
}

TEST (SDP_Tests, SDP_Test_read_product_identifier) {
    int16_t error;
    uint32_t product_number;
    uint8_t serial_number[8];
    uint8_t serial_number_size = 8;
    error = sdp_prepare_product_identifier();
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_prepare_product_identifier");
    error = sdp_read_product_identifier(&product_number, &serial_number[0],
                                        serial_number_size);
    CHECK_EQUAL_ZERO_TEXT(error, "sdp_read_product_identifier");
    printf("Product number: %u\n", product_number);
    printf("Serial number: ");
    for (size_t i = 0; i < serial_number_size; i++) {
        printf("%u, ", serial_number[i]);
    }
    printf("\n");
}


sdp_i2c.c
/*
 * THIS FILE IS AUTOMATICALLY GENERATED
 *
 * I2C-Generator: 0.2.0
 * Yaml Version: 0.1.0
 * Template Version: 0.7.0-32-g5b8f0ff
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "sdp_i2c.h"
#include "sensirion_common.h"
#include "sensirion_i2c.h"
#include "sensirion_i2c_hal.h"

int16_t
sdp_start_continuous_measurement_with_mass_flow_t_comp_and_averaging(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x3603);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(10000);
    return NO_ERROR;
}

int16_t sdp_start_continuous_measurement_with_mass_flow_t_comp(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x3608);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(10000);
    return NO_ERROR;
}

int16_t
sdp_start_continuous_measurement_with_diff_pressure_t_comp_and_averaging(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x3615);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(10000);
    return NO_ERROR;
}

int16_t sdp_start_continuous_measurement_with_diff_pressure_t_comp(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x361E);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(10000);
    return NO_ERROR;
}

int16_t sdp_stop_continuous_measurement(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x3FF9);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(500);
    return NO_ERROR;
}

int16_t sdp_trigger_measurement_with_mass_flow_t_comp_and_averaging(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x3624);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(45000);
    return NO_ERROR;
}

int16_t sdp_trigger_measurement_with_diff_pressure_t_comp(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x362F);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }
    sensirion_i2c_hal_sleep_usec(45000);
    return NO_ERROR;
}

int16_t sdp_read_measurement(int16_t* differential_pressure,
                             int16_t* temperature, int16_t* scaling_factor) {
    int16_t error;
    uint8_t buffer[9];
    error = sensirion_i2c_read_data_inplace(SDP_I2C_ADDRESS, &buffer[0], 6);
    if (error) {
        return error;
    }
    *differential_pressure = sensirion_common_bytes_to_int16_t(&buffer[0]);
    *temperature = sensirion_common_bytes_to_int16_t(&buffer[2]);
    *scaling_factor = sensirion_common_bytes_to_int16_t(&buffer[4]);
    return NO_ERROR;
}

int16_t sdp_enter_sleep_mode(void) {
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x3677);

    return sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
}

int16_t sdp_exit_sleep_mode(void) {
    int16_t error;
    uint8_t buffer[2];
    uint16_t offset = 0;
    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    sensirion_i2c_hal_sleep_usec(2000);
    return error;
}

int16_t sdp_prepare_product_identifier(void) {
    uint8_t buffer[2];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0x367C);

    return sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
}

int16_t sdp_read_product_identifier(uint32_t* product_number,
                                    uint8_t* serial_number,
                                    uint8_t serial_number_size) {
    int16_t error;
    uint8_t buffer[18];
    uint16_t offset = 0;
    offset = sensirion_i2c_add_command_to_buffer(&buffer[0], offset, 0xE102);

    error = sensirion_i2c_write_data(SDP_I2C_ADDRESS, &buffer[0], offset);
    if (error) {
        return error;
    }

    sensirion_i2c_hal_sleep_usec(1000);

    error = sensirion_i2c_read_data_inplace(SDP_I2C_ADDRESS, &buffer[0], 12);
    if (error) {
        return error;
    }
    *product_number = sensirion_common_bytes_to_uint32_t(&buffer[0]);
    sensirion_common_copy_bytes(&buffer[4], serial_number, serial_number_size);
    return NO_ERROR;
}


sdp_i2c.h
/*
 * THIS FILE IS AUTOMATICALLY GENERATED
 *
 * I2C-Generator: 0.2.0
 * Yaml Version: 0.1.0
 * Template Version: 0.7.0-32-g5b8f0ff
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SDP_I2C_H
#define SDP_I2C_H

#ifdef __cplusplus
extern "C" {
#endif

#include "sensirion_config.h"

#define SDP8XX_I2C_ADDRESS_0 0x25
#define SDP8XX_I2C_ADDRESS_1 0x26
#define SDP3X_I2C_ADDRESS_0 0x21
#define SDP3X_I2C_ADDRESS_1 0x22
#define SDP3X_I2C_ADDRESS_2 0x23

#define SDP_I2C_ADDRESS SDP8XX_I2C_ADDRESS_0

/**
 * sdp_start_continuous_measurement_with_mass_flow_t_comp_and_averaging() - This
 * command starts continuous measurements with mass flow temperature
 * compensation and the average till read feature.
 *
 * @note The measurement command must only be sent once, if acknowledged. The
 * command must not be resent or other commands must not be sent until the stop
 * measurement command has been issued. After the start measurement command is
 * sent, the first measurement result is available after 8ms. Small accuracy
 * deviations (few % of reading) can occur during the next 12ms. The measured
 * values are updated every 0.5ms and can be read using the read measurement
 * interface.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t
sdp_start_continuous_measurement_with_mass_flow_t_comp_and_averaging(void);

/**
 * sdp_start_continuous_measurement_with_mass_flow_t_comp() - This command
 * starts continuous measurements with mass flow temperature compensation.
 *
 * @note The measurement command must only be sent once, if acknowledged. The
 * command must not be resent or other commands must not be sent until the stop
 * measurement command has been issued. After the start measurement command is
 * sent, the first measurement result is available after 8ms. Small accuracy
 * deviations (few % of reading) can occur during the next 12ms. The measured
 * values are updated every 0.5ms and can be read using the read measurement
 * interface.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_start_continuous_measurement_with_mass_flow_t_comp(void);

/**
 * sdp_start_continuous_measurement_with_diff_pressure_t_comp_and_averaging() -
 * This command starts continuous measurements with differential pressure
 * temperature compensation and the average till read feature.
 *
 * @note The measurement command must only be sent once, if acknowledged. The
 * command must not be resent or other commands must not be sent until the stop
 * measurement command has been issued. After the start measurement command is
 * sent, the first measurement result is available after 8ms. Small accuracy
 * deviations (few % of reading) can occur during the next 12ms. The measured
 * values are updated every 0.5ms and can be read using the read measurement
 * interface.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t
sdp_start_continuous_measurement_with_diff_pressure_t_comp_and_averaging(void);

/**
 * sdp_start_continuous_measurement_with_diff_pressure_t_comp() - This command
 * starts continuous measurements with differential pressure temperature
 * compensation.
 *
 * @note The measurement command must only be sent once, if acknowledged. The
 * command must not be resent or other commands must not be sent until the stop
 * measurement command has been issued. After the start measurement command is
 * sent, the first measurement result is available after 8ms. Small accuracy
 * deviations (few % of reading) can occur during the next 12ms. The measured
 * values are updated every 0.5ms and can be read using the read measurement
 * interface.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_start_continuous_measurement_with_diff_pressure_t_comp(void);

/**
 * sdp_stop_continuous_measurement() - This command stops the continuous
 * measurement and puts the sensor in idle mode. It powers off the heater and
 * makes the sensor receptive to another command after 500us. The Stop command
 * is also required when switching between different continuous measurement
 * commands.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_stop_continuous_measurement(void);

/**
 * sdp_trigger_measurement_with_mass_flow_t_comp_and_averaging() - This command
 * triggers a single shot measurement with mass flow temperature compensation.
 *
 * @note During a triggered measurement the sensor measures both differential
 * pressure and temperature. The measurement starts directly after the command
 * has been sent. The command needs to be repeated with every measurement.
 * During the 45ms that the sensor is measuring, no command can be sent to the
 * sensor. After the 45ms the result can be read out and any command can be sent
 * to the sensor.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_trigger_measurement_with_mass_flow_t_comp_and_averaging(void);

/**
 * sdp_trigger_measurement_with_diff_pressure_t_comp() - This command triggers a
 * single shot measurement with differential pressure temperature compensation.
 *
 * @note During a triggered measurement the sensor measures both differential
 * pressure and temperature. The measurement starts directly after the command
 * has been sent. The command needs to be repeated with every measurement.
 * During the 45ms that the sensor is measuring, no command can be sent to the
 * sensor. After the 45ms the result can be read out and any command can be sent
 * to the sensor.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_trigger_measurement_with_diff_pressure_t_comp(void);

/**
 * sdp_read_measurement() - After a start continuous measurement commands, the
 * measurement results can be read out at most every 0.5ms. After a triggered
 * measurement command, the results can be read out when the sensor is finished
 * with the measurement. The temperature and scale factor don’t need to be read
 * out (every time). The read sequence can be aborted by a NACK and a STOP
 * condition. The scale factor is for differential pressure in Pascal.
 *
 * @param differential_pressure The digital calibrated differential pressure
 * signal read from the sensor is a signed integer number (two's complement
 * number). The integer value can be converted to the physical value by dividing
 * it by the scale factor. differential pressure in Pascal = sensor output /
 * scale factor
 *
 * @param temperature The digital calibrated temperature signal read from the
 * sensor is a signed integer number (two's complement number). The integer
 * value can be converted to the physical value by dividing it by a scale factor
 * of 200. temperature in °C = sensor output / 200
 *
 * @param scaling_factor Scaling factor differential pressure
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_read_measurement(int16_t* differential_pressure,
                             int16_t* temperature, int16_t* scaling_factor);

/**
 * sdp_enter_sleep_mode() - In sleep mode the sensor uses the minimum amount of
 * current. The mode can only be entered from idle mode, i.e. when the sensor is
 * not measuring. This mode is particularly useful for battery operated devices.
 * To minimize the current in this mode, the complexity of the sleep mode
 * circuit has been reduced as much as possible, which is mainly reflected by
 * the way the sensor exits the sleep mode. In sleep mode the sensor cannot be
 * soft reset.
 *
 * @note Triggered mode: the sleep command can be sent after the result have
 * been read out and the sensor is in idle mode. Continuous mode: the sleep
 * command can be sent after a stop continuous measurement command has been
 * issued and the sensor is in idle mode.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_enter_sleep_mode(void);

/**
 * sdp_exit_sleep_mode() - Exit sleep mode. See the data sheet for more detailed
 * information
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_exit_sleep_mode(void);

/**
 * sdp_prepare_product_identifier() - Prepare for reading the product identifier
 * and sensor serial number.
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_prepare_product_identifier(void);

/**
 * sdp_read_product_identifier() - Read the product identifier and serial number
 * of the sensor.
 *
 * @note Make sure to call 'prepare product identifier' immediately before.
 *
 * @param product_number 32 bit unique product and revision number. The number
 * is listed in the datasheet. Note that the last 8 bits are the revision number
 * and can be subject to change.
 *
 * @param serial_number 64-bit unique serial number
 *
 * @return 0 on success, an error code otherwise
 */
int16_t sdp_read_product_identifier(uint32_t* product_number,
                                    uint8_t* serial_number,
                                    uint8_t serial_number_size);

#ifdef __cplusplus
}
#endif

#endif /* SDP_I2C_H */


sdp_i2c_example_usage.c
/*
 * I2C-Generator: 0.2.0
 * Yaml Version: 0.1.0
 * Template Version: 0.7.0-32-g5b8f0ff
 */
/*
 * Copyright (c) 2021, Sensirion AG
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of Sensirion AG nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>  // printf

#include "sdp_i2c.h"
#include "sensirion_common.h"
#include "sensirion_i2c_hal.h"
#include <inttypes.h>

/*
 * TO USE CONSOLE OUTPUT (PRINTF) YOU MAY NEED TO ADAPT THE INCLUDE ABOVE OR
 * DEFINE IT ACCORDING TO YOUR PLATFORM:
 * #define printf(...)
 */

int main(void) {
    int16_t error = 0;

    sensirion_i2c_hal_init();

    // Change the used i2c address 'SDP_I2C_ADDRESS' in sdp_i2c.h
    printf("Launching SDP driver with I2C address: 0x%x\n", SDP_I2C_ADDRESS);

    sdp_stop_continuous_measurement();
    // ignore error if no measurement running

    uint32_t product_number;
    uint8_t serial[8];
    uint8_t serial_size = 8;

    error = sdp_prepare_product_identifier();
    if (error) {
        printf("Error executing sdp_prepare_product_identifier(): %i\n", error);
    }

    error = sdp_read_product_identifier(&product_number, serial, serial_size);

    if (error) {
        printf("Error executing sdp_read_product_identifier(): %i\n", error);
    } else {
        // uint64_t serial_number =
        //    (uint64_t)serial[0] << 56 | (uint64_t)serial[1] << 48 |
        //    (uint64_t)serial[2] << 40 | (uint64_t)serial[3] << 32 |
        //    (uint64_t)serial[4] << 24 | (uint64_t)serial[5] << 16 |
        //    (uint64_t)serial[6] << 8 | (uint64_t)serial[7];
        printf("Product Number: 0x%08x\n", product_number);
        // printf("Serial Number: %" PRIu64 "\n", serial_number);
    }

    // Start Measurement
    error =
        sdp_start_continuous_measurement_with_diff_pressure_t_comp_and_averaging();
    if (error) {
        printf("Error executing "
               "sdp_start_continuous_measurement_with_diff_pressure_t_comp_"
               "and_averaging(): %i\n",
               error);
    }

    for (int i = 0; i < 10000; i++) {
        // Read Measurement
        sensirion_i2c_hal_sleep_usec(10000);

        int16_t differential_pressure;
        int16_t temperature;
        int16_t scaling_factor;

        error = sdp_read_measurement(&differential_pressure, &temperature,
                                     &scaling_factor);

        if (error) {
            printf("Error executing sdp_read_measurement(): %i\n", error);
        } else {
            printf("Differential pressure: %i\n", differential_pressure);
            printf("Temperature: %i\n", temperature);
            printf("Scaling factor: %i\n", scaling_factor);
        }
    }

    error = sdp_stop_continuous_measurement();
    if (error) {
        printf("Error executing sdp_stop_continuous_measurement(): %i\n",
               error);
    }

    return 0;
}